tests/h/services/bulk_annotation/service_test.py

```diff
@@ -0,0 +1,87 @@
+import pytest
+
+from h.services.bulk_annotation import BulkAnnotationService
+
+
+class TestBulkAnnotationService:
+    AUTHORITY = "my.authority"
+
+    @pytest.mark.parametrize(
+        "key,value,visible",
+        (
+            (None, None, True),
+            ("shared", False, False),
+            ("deleted", True, False),
+            ("nipaed", True, False),
+            ("moderated", True, False),
+            ("updated", "2020-01-01", False),
+            ("updated", "2020-01-02", True),
+            ("updated", "2022-01-01", True),
+            ("updated", "2022-01-02", False),
+        ),
+    )
+    def test_it_with_single_annotation(self, svc, factories, key, value, visible):
+        values = {
+            "shared": True,
+            "deleted": False,
+            "nipaed": False,
+            "moderated": False,
+            "updated": "2021-01-01",
+        }
+        if key:
+            values[key] = value
+
+        viewer = factories.User(authority=self.AUTHORITY)
+        author = factories.User(authority=self.AUTHORITY, nipsa=values["nipaed"])
+        group = factories.Group(members=[author, viewer])
+        anno = factories.Annotation(
+            userid=author.userid,
+            group=group,
+            shared=values["shared"],
+            deleted=values["deleted"],
+            updated=values["updated"],
+        )
+        if values["moderated"]:
+            factories.AnnotationModeration(annotation=anno)
+
+        annotations = svc.annotation_search(
+            authority=self.AUTHORITY,
+            audience={"username": [viewer.username]},
+            updated={"gt": "2020-01-01", "lte": "2022-01-01"},
+        )
+
+        if visible:
+            assert annotations == [anno]
+        else:
+            assert not annotations
+
+    def test_it_with_more_complex_grouping(self, svc, factories):
+        viewer_1, viewer_2, author = factories.User.create_batch(
+            3, authority=self.AUTHORITY
+        )
+
+        relevant_group_1 = factories.Group(members=[viewer_1, author])
+        relevant_group_2 = factories.Group(members=[viewer_2, author])
+        irrelevant_group = factories.Group(members=[author])
+
+        visible_anno_1 = factories.Annotation(
+            userid=author.userid, group=relevant_group_1, shared=True, deleted=False
+        )
+        visible_anno_2 = factories.Annotation(
+            userid=author.userid, group=relevant_group_2, shared=True, deleted=False
+        )
+        _invisible_anno = factories.Annotation(
+            userid=author.userid, group=irrelevant_group, shared=True, deleted=False
+        )
```

Comment:
I wonder if "matching" and "non_matching" might service instead of relevant/irrelevant and visible/invisible.

Also wonder if the users, groups and annotations might work nicely as lists. Then `viewer_1` and `viewer_2` quite nicely become `viewers[0]` and `viewers[1]`, and the same for `matching_groups` and `matching_annotations`:

```suggestion
        *viewers, author = factories.User.create_batch(3, authority=self.AUTHORITY)

        matching_groups = [
            factories.Group(members=[viewer, author])
            for viewer in viewers
        ]
        non_matching_group = factories.Group(members=[author])

        matching_annotations = [
            factories.Annotation(
                userid=author.userid, group=group, shared=True, deleted=False
            )
            for group in matching_groups
        ]
        non_matching_annotation = factories.Annotation(
            userid=author.userid, group=non_matching_group, shared=True, deleted=False
        )
```

Reply:
I've tried this out. Are list comprehensions the good kind of clever here? As an example, at a glance which users are in which groups in these two layouts:

```python
*viewers, author = factories.User.create_batch(3, authority=self.AUTHORITY)

matching_groups = [
    factories.Group(members=[viewer, author])
    for viewer in viewers
]
non_matching_group = factories.Group(members=[author])
```

```python
viewer_1, viewer_2, author = factories.User.create_batch(
    3, authority=self.AUTHORITY
)

relevant_group_1 = factories.Group(members=[viewer_1, author])
relevant_group_2 = factories.Group(members=[viewer_2, author])
irrelevant_group = factories.Group(members=[author])
```

Reply:
I've come up with a version that does everything at once in one comprehension. This might be clearer?

```python
        *viewers, author = factories.User.create_batch(3, authority=self.AUTHORITY)

        annotations = [
            factories.Annotation(
                userid=author.userid,
                group=factories.Group(members=group_members),
                shared=True,
                deleted=False,
            )
            for group_members in (
                # The first two annotations should match, because they are in
                # groups the viewers are in
                [author, viewers[0]],
                [author, viewers[1]],
                # This one is just noise and shouldn't match
                [author],
            )
        ]
```